### PR TITLE
8009-handling-shipping-adress: REDUX

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
:+1:

### CODE REVIEW

File: app/helpers/checkout_helper.rb [hunk 1]
```diff
@@ -1,6 +1,10 @@
 # frozen_string_literal: true
 
 module CheckoutHelper
+  def shipping_and_billing_match?(order)
+    order.ship_address == order.bill_address
+  end
+
```
Comment (andrewpbrett) [inline app/helpers/checkout_helper.rb:7]: This doesn't actually do the same thing as the extracted method; if the two addresses have all the same attributes but different ID's, this returns false. In that scenario I think we should consider them matching addresses.

File: app/controllers/split_checkout_controller.rb [hunk 2]
```diff
@@ -15,146 +15,55 @@ class SplitCheckoutController < ::BaseController
...
-      :state_id,
-      :zipcode,
-      :country_id,
-      :id
-    ]
-    address_attrs.each do |attr|
-      next if params[:order][:ship_address_attributes][attr].present?
-
-      params[:order][:ship_address_attributes][attr] = params[:order][:bill_address_attributes][attr]
-    end
```
Comment (andrewpbrett) [inline app/controllers/split_checkout_controller.rb:102]: This was deleted but I don't think the problem it was solving is accounted for in any of the new code.

**Review by andrewpbrett — CHANGES_REQUESTED** [review]
Thanks @matt-yorkley, I went through and added some questions on some of the refactoring.

File: app/controllers/split_checkout_controller.rb [hunk 3]
```diff
@@ -91,33 +78,4 @@ def redirect_to_step
...
-  def handle_redirect_from_stripe
-    return checkout_failed unless @order.process_payments!
-
-    if OrderWorkflow.new(@order).next && order_complete?
-      checkout_succeeded
-      redirect_to(order_path(@order)) && return
-    else
-      checkout_failed
-    end
-  end
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/controllers/split_checkout_controller.rb:191]: None of it is _currently_ being used by the split checkout, and when we get to implementing the payment handling, the code will probably look a bit different.

File: app/controllers/split_checkout_controller.rb [hunk 4]
```diff
@@ -15,146 +15,55 @@ class SplitCheckoutController < ::BaseController
...
 
     OrderWorkflow.new(@order).next if @order.cart?
-
-    # This is only required because of spree_paypal_express. If we implement
-    # a version of paypal that uses this controller, and more specifically
-    # the #action_failed method, then we can remove this call
-    # OrderCheckoutRestart.new(@order).call
-    @order.errors.clear
-    @order.bill_address.errors.clear
-    @order.ship_address.errors.clear
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/controllers/split_checkout_controller.rb:30]: Ah, I was using a logged-in user with a saved address. Updated :+1:

File: app/helpers/checkout_helper.rb (hunk 1 above)
Comment (Matt-Yorkley) [reply app/helpers/checkout_helper.rb:7]: > if the two addresses have all the same attributes but different ID's, this returns false

The `Spree::Address` class has it's own implementation of the `==` method, so it should return `true` in that scenario:

https://github.com/openfoodfoundation/openfoodnetwork/blob/3e0509d23241dd0ae71f02f1f221a78ef03a0b3b/app/models/spree/address.rb#L62-L71

Comment (andrewpbrett) [reply app/helpers/checkout_helper.rb:7]: Ahh nice, good catch. That makes this way cleaner 👍

File: app/controllers/split_checkout_controller.rb (hunk 2 above)
Comment (Matt-Yorkley) [reply app/controllers/split_checkout_controller.rb:102]: We should start adding some tests soon to verify this stuff. What problem was it solving? We need to copy `:firstname`, `:lastname`, and `:phone` from `bill_address` to the `ship_address`. That's being done :ok_hand:

**Comment by andrewpbrett** [comment]
This is looking good now, thanks @Matt-Yorkley. @jibees will have to be the one to merge it to his branch tomorrow. I think once we do that we'll be pretty close to being able to merge https://github.com/openfoodfoundation/openfoodnetwork/pull/8076

Once that's merged, we can rebase and continue the work @jibees started in https://github.com/openfoodfoundation/openfoodnetwork/pull/8070 and add additional tests.

**Pull request merged**